File: src/test/java/org/javacs/kt/OneFilePerformance.java
package org.javacs.kt;

import com.intellij.openapi.Disposable;
import com.intellij.openapi.vfs.StandardFileSystems;
import com.intellij.openapi.vfs.VirtualFile;
import com.intellij.openapi.vfs.VirtualFileManager;
import com.intellij.openapi.vfs.VirtualFileSystem;
import com.intellij.psi.PsiManager;
import com.intellij.psi.util.PsiTreeUtil;
import org.jetbrains.kotlin.cli.jvm.compiler.CliBindingTrace;
import org.jetbrains.kotlin.cli.jvm.compiler.EnvironmentConfigFiles;
import org.jetbrains.kotlin.cli.jvm.compiler.KotlinCoreEnvironment;
import org.jetbrains.kotlin.config.CommonConfigurationKeys;
import org.jetbrains.kotlin.config.CompilerConfiguration;
import org.jetbrains.kotlin.container.ComponentProvider;
import org.jetbrains.kotlin.container.ValueDescriptor;
import org.jetbrains.kotlin.descriptors.CallableDescriptor;
import org.jetbrains.kotlin.load.java.JvmAbi;
import org.jetbrains.kotlin.psi.KtElement;
import org.jetbrains.kotlin.psi.KtFile;
import org.jetbrains.kotlin.psi.KtPsiFactory;
import org.jetbrains.kotlin.resolve.BindingTraceContext;
import org.jetbrains.kotlin.resolve.LazyTopDownAnalyzer;
import org.jetbrains.kotlin.resolve.TopDownAnalysisMode;
import org.jetbrains.kotlin.resolve.calls.callUtil.CallUtilKt;
import org.jetbrains.kotlin.resolve.calls.model.ResolvedCall;
import org.jetbrains.kotlin.resolve.calls.smartcasts.DataFlowInfoFactory;
import org.junit.Test;
import org.openjdk.jmh.annotations.Benchmark;
import org.openjdk.jmh.annotations.Scope;
import org.openjdk.jmh.annotations.State;
import org.openjdk.jmh.runner.Runner;
import org.openjdk.jmh.runner.RunnerException;
import org.openjdk.jmh.runner.options.Options;
import org.openjdk.jmh.runner.options.OptionsBuilder;

import java.lang.reflect.Type;
import java.net.URL;
import java.util.Arrays;
import java.util.Collection;
import java.util.Collections;

public class OneFilePerformance {
    @State(Scope.Thread)
    public static class ReusableParts {
        CompilerConfiguration config = new CompilerConfiguration();
        {
            config.put(CommonConfigurationKeys.MODULE_NAME, JvmAbi.DEFAULT_MODULE_NAME);
        }
        KotlinCoreEnvironment env = KotlinCoreEnvironment.createForProduction(new Disposable() {
            @Override
            public void dispose() {
                // Do nothing
            }
        }, config, EnvironmentConfigFiles.JVM_CONFIG_FILES);
        KtPsiFactory parser = new KtPsiFactory(env.getProject());
        VirtualFileSystem fileSystem = VirtualFileManager.getInstance().getFileSystem(StandardFileSystems.FILE_PROTOCOL);
        KtFile bigFile = openFile("/kotlinCompilerPerformance/BigFile.kt");

        KtFile openFile(String resourcePath) {
            URL locate = OneFilePerformance.class.getResource(resourcePath);
            VirtualFile file = fileSystem.findFileByPath(locate.getPath());
            return (KtFile) PsiManager.getInstance(env.getProject()).findFile(file);
        }

        BindingTraceContext compile(Collection<KtFile> compile, Collection<KtFile> sourcePath) {
            BindingTraceContext trace = new CliBindingTrace();
            new ComponentProvider(){

                @Override
                public ValueDescriptor resolve(Type arg0) {
                    return null;
                }

                @Override
                public <T> T create(Class<T> arg0) {
                    return null;
                }
            };

            // CompilerFixtures is a Kotlin class in this package
            // and thus might generate an error when using a Java linter.
            // It is safe to ignore that error.
            ComponentProvider container = CompilerFixtures.createContainer(
                    env.getProject(),
                    sourcePath,
                    trace,
                    env.getConfiguration(),
                    env::createPackagePartProvider);
            LazyTopDownAnalyzer analyze = container.create(LazyTopDownAnalyzer.class);

            analyze.analyzeDeclarations(TopDownAnalysisMode.TopLevelDeclarations, compile, DataFlowInfoFactory.EMPTY);

            return trace;
        }
    }

    @Benchmark
    public void recompileBigFile(ReusableParts state) {
        state.compile(Collections.singletonList(state.bigFile), Collections.singletonList(state.bigFile));
    }

    @Benchmark
    public void recompileSmallFile(ReusableParts state) {
        KtFile smallFile = state.openFile("/kotlinCompilerPerformance/ReferencesBigFile.kt");
        BindingTraceContext analyze = state.compile(Arrays.asList(smallFile), Arrays.asList(smallFile, state.bigFile));

        // Check reference
        KtElement ref = PsiTreeUtil.getNonStrictParentOfType(smallFile.findElementAt(80), KtElement.class);
        ResolvedCall<? extends CallableDescriptor> call = CallUtilKt.getParentResolvedCall(ref, analyze.getBindingContext(), false);
        if (!call.getCandidateDescriptor().getName().asString().equals("max"))
            throw new RuntimeException("Expected BigFile.max but found " + call);
    }

    @Benchmark
    public void recompileBoth(ReusableParts state) {
        KtFile bigFile = state.openFile("/kotlinCompilerPerformance/BigFile.kt");
        KtFile smallFile = state.openFile("/kotlinCompilerPerformance/ReferencesBigFile.kt");
        BindingTraceContext analyze = state.compile(Arrays.asList(smallFile, bigFile), Arrays.asList(smallFile, bigFile));

        // Check reference
        KtElement ref = PsiTreeUtil.getNonStrictParentOfType(smallFile.findElementAt(80), KtElement.class);
        ResolvedCall<? extends CallableDescriptor> call = CallUtilKt.getParentResolvedCall(ref, analyze.getBindingContext(), false);
        if (!call.getCandidateDescriptor().getName().asString().equals("max"))
            throw new RuntimeException("Expected BigFile.max but found " + call);
    }

    @Test
    public void checkRecompileBoth() {
        ReusableParts state = new ReusableParts();
        recompileBoth(state);
    }

    @Test
    public void checkRecompileOne() {
        ReusableParts state = new ReusableParts();
        state.compile(Collections.singleton(state.bigFile), Collections.singleton(state.bigFile));
        recompileSmallFile(state);
    }

    public static void main(String[] args) throws RunnerException, InterruptedException {
        Options opt = new OptionsBuilder().include(OneFilePerformance.class.getSimpleName())
                .forks(1)
                .threads(1)
                .warmupIterations(5)
                .measurementIterations(5)
                .build();
        new Runner(opt).run();
    }
}

File: server/src/test/kotlin/org/javacs/kt/OneFilePerformance.kt
package org.javacs.kt

import org.javacs.kt.util.LoggingMessageCollector
import org.jetbrains.kotlin.cli.common.CLIConfigurationKeys
import org.jetbrains.kotlin.cli.jvm.compiler.CliBindingTrace
import org.jetbrains.kotlin.cli.jvm.compiler.EnvironmentConfigFiles
import org.jetbrains.kotlin.cli.jvm.compiler.KotlinCoreEnvironment
import org.jetbrains.kotlin.com.intellij.openapi.util.Disposer
import org.jetbrains.kotlin.com.intellij.openapi.vfs.StandardFileSystems
import org.jetbrains.kotlin.com.intellij.openapi.vfs.VirtualFileManager
import org.jetbrains.kotlin.com.intellij.psi.PsiManager
import org.jetbrains.kotlin.com.intellij.psi.util.PsiTreeUtil
import org.jetbrains.kotlin.config.CommonConfigurationKeys
import org.jetbrains.kotlin.config.CompilerConfiguration
import org.jetbrains.kotlin.metadata.jvm.deserialization.JvmProtoBufUtil
import org.jetbrains.kotlin.psi.KtElement
import org.jetbrains.kotlin.psi.KtFile
import org.jetbrains.kotlin.psi.KtPsiFactory
import org.jetbrains.kotlin.resolve.BindingTraceContext
import org.jetbrains.kotlin.resolve.LazyTopDownAnalyzer
import org.jetbrains.kotlin.resolve.TopDownAnalysisMode
import org.jetbrains.kotlin.resolve.calls.callUtil.getParentResolvedCall
import org.jetbrains.kotlin.resolve.calls.smartcasts.DataFlowInfoFactory
import org.junit.Test
import org.openjdk.jmh.annotations.*
import org.openjdk.jmh.runner.Runner
import org.openjdk.jmh.runner.RunnerException
import org.openjdk.jmh.runner.options.OptionsBuilder
import java.io.Closeable
import java.net.URLDecoder
import java.nio.charset.StandardCharsets

class OneFilePerformance {
    @State(Scope.Thread)
    class ReusableParts : Closeable {
        internal var config = CompilerConfiguration()
        init {
            config.put(CommonConfigurationKeys.MODULE_NAME, JvmProtoBufUtil.DEFAULT_MODULE_NAME)
            config.put(CLIConfigurationKeys.MESSAGE_COLLECTOR_KEY, LoggingMessageCollector)
        }
        internal val disposable = Disposer.newDisposable()
        internal var env = KotlinCoreEnvironment.createForProduction(
            disposable, config, EnvironmentConfigFiles.JVM_CONFIG_FILES
        )
        internal var parser = KtPsiFactory(env.project)
        internal var fileSystem = VirtualFileManager.getInstance().getFileSystem(StandardFileSystems.FILE_PROTOCOL)
        internal var bigFile = openFile("/kotlinCompilerPerformance/BigFile.kt")

        internal fun openFile(resourcePath: String?): KtFile {
            val locate = OneFilePerformance::class.java.getResource(resourcePath)
            val file = fileSystem.findFileByPath(URLDecoder.decode(locate.path, StandardCharsets.UTF_8.toString()))
            return PsiManager.getInstance(env.project).findFile(file!!) as KtFile
        }

        internal fun compile(compile: Collection<KtFile>, sourcePath: Collection<KtFile>): BindingTraceContext {
            val trace = CliBindingTrace()
            val container = CompilerFixtures.createContainer(
                env.project,
                sourcePath,
                trace,
                env.configuration,
                env::createPackagePartProvider
            )
            val analyze = container.create(LazyTopDownAnalyzer::class.java)
            analyze.analyzeDeclarations(TopDownAnalysisMode.TopLevelDeclarations, compile, DataFlowInfoFactory.EMPTY)
            return trace
        }

        @TearDown(Level.Trial)
        override fun close() {
            Disposer.dispose(disposable)
        }
    }

    @Benchmark
    fun recompileBigFile(state: ReusableParts) {
        val path = listOf(state.bigFile)
        state.compile(path, path)
    }

    @Benchmark
    fun recompileSmallFile(state: ReusableParts) {
        val smallFile = state.openFile("/kotlinCompilerPerformance/ReferencesBigFile.kt")
        val analyze = state.compile(listOf(smallFile), listOf(smallFile, state.bigFile))
        // Check reference
        val ref = PsiTreeUtil.getNonStrictParentOfType(smallFile.findElementAt(80), KtElement::class.java)
        val call = ref.getParentResolvedCall(analyze.getBindingContext(), false)
        if (!call!!.getCandidateDescriptor().getName().asString().equals("max"))
            throw RuntimeException("Expected BigFile.max but found " + call)
    }

    @Benchmark
    fun recompileBoth(state: ReusableParts) {
        val bigFile = state.openFile("/kotlinCompilerPerformance/BigFile.kt")
        val smallFile = state.openFile("/kotlinCompilerPerformance/ReferencesBigFile.kt")
        val analyze = state.compile(listOf(smallFile, bigFile), listOf(smallFile, bigFile))
        // Check reference
        val ref = PsiTreeUtil.getNonStrictParentOfType(smallFile.findElementAt(80), KtElement::class.java)
        val call = ref.getParentResolvedCall(analyze.getBindingContext(), false)
        if (!call!!.getCandidateDescriptor().getName().asString().equals("max"))
            throw RuntimeException("Expected BigFile.max but found " + call)
    }

    @Test
    fun checkRecompileBoth() {
        ReusableParts().use { state ->
            recompileBoth(state)
        }
    }

    @Test
    fun checkRecompileOne() {
        ReusableParts().use { state ->
            state.compile(setOf(state.bigFile), setOf(state.bigFile))
            recompileSmallFile(state)
        }
    }

    companion object {
        @Throws(RunnerException::class, InterruptedException::class)
        @JvmStatic fun main(args: Array<String>) {
            val opt = OptionsBuilder().include(OneFilePerformance::class.java.getSimpleName())
                    .forks(1)
                    .threads(1)
                    .warmupIterations(5)
                    .measurementIterations(5)
                    .build()
            Runner(opt).run()
        }
    }
}
